### PR TITLE
Closes #37

### LINKED ISSUE #37: Create dataset loader for n2c2 2018 - Adverse Drug Events and Medication Extraction
From https://portal.dbmi.hms.harvard.edu/projects/n2c2-nlp/

### PR DESCRIPTION
**Pull request by singhay**
Please name your PR after the issue it closes. You can use the following line: "Closes #ISSUE-NUMBER" where you replace the ISSUE-NUMBER with the one corresponding to your dataset.

If the following information is NOT present in the issue, please populate:

- **Name:** *name of the dataset*
- **Description:** *short description of the dataset (or link to social media or blog post)*
- **Paper:** *link to the dataset paper if available*
- **Data:** *link to the online home of the dataset*

### Checkbox

- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `biodatasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_BIGBIO_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `BigBioConfig` for the source schema and one for a bigbio schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_bigbio biodatasets/my_dataset/my_dataset.py`.
- [x] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

```
❯ python -m tests.test_bigbio biodatasets/n2c2_2018/n2c2_2018_track2.py --data_dir /Users/ayush.singh/workspace/data/Healthcare/i2b2_2018_ADE_Medication
INFO:__main__:args: Namespace(path='biodatasets/n2c2_2018/n2c2_2018_track2.py', schema=None, subset_id=None, data_dir='/Users/ayush.singh/workspace/data/Healthcare/i2b2_2018_ADE_Medication', use_auth_token=None)
INFO:__main__:self.PATH: biodatasets/n2c2_2018/n2c2_2018_track2.py
INFO:__main__:self.SUBSET_ID: n2c2_2018_track2
INFO:__main__:self.SCHEMA: None
INFO:__main__:self.DATA_DIR: /Users/ayush.singh/workspace/data/Healthcare/i2b2_2018_ADE_Medication
INFO:__main__:Checking for _SUPPORTED_TASKS ...
INFO:__main__:Found _SUPPORTED_TASKS=[<Tasks.NAMED_ENTITY_RECOGNITION: 'NER'>, <Tasks.RELATION_EXTRACTION: 'RE'>]
INFO:__main__:_SUPPORTED_TASKS implies _MAPPED_SCHEMAS={'KB'}
INFO:__main__:schemas_to_check: {'KB'}
INFO:__main__:Checking load_dataset with config name n2c2_2018_track2_source
WARNING:datasets.builder:Using custom data configuration n2c2_2018_track2_source-069ba31903658442
Downloading and preparing dataset n2_c2_adverse_drug_events_medication_extraction_dataset/n2c2_2018_track2_source to /Users/ayush.singh/.cache/huggingface/datasets/n2_c2_adverse_drug_events_medication_extraction_dataset/n2c2_2018_track2_source-069ba31903658442/1.0.0/391e9ca2838fa532d14581d6ba6a63062b6fb22b50033f153ab9ddf0c4780cb7...
Dataset n2_c2_adverse_drug_events_medication_extraction_dataset downloaded and prepared to /Users/ayush.singh/.cache/huggingface/datasets/n2_c2_adverse_drug_events_medication_extraction_dataset/n2c2_2018_track2_source-069ba31903658442/1.0.0/391e9ca2838fa532d14581d6ba6a63062b6fb22b50033f153ab9ddf0c4780cb7. Subsequent calls will reuse this data.
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 187.97it/s]
INFO:__main__:Checking load_dataset with config name n2c2_2018_track2_bigbio_kb
WARNING:datasets.builder:Using custom data configuration n2c2_2018_track2_bigbio_kb-069ba31903658442
Downloading and preparing dataset n2_c2_adverse_drug_events_medication_extraction_dataset/n2c2_2018_track2_bigbio_kb to /Users/ayush.singh/.cache/huggingface/datasets/n2_c2_adverse_drug_events_medication_extraction_dataset/n2c2_2018_track2_bigbio_kb-069ba31903658442/1.0.0/391e9ca2838fa532d14581d6ba6a63062b6fb22b50033f153ab9ddf0c4780cb7...
Dataset n2_c2_adverse_drug_events_medication_extraction_dataset downloaded and prepared to /Users/ayush.singh/.cache/huggingface/datasets/n2_c2_adverse_drug_events_medication_extraction_dataset/n2c2_2018_track2_bigbio_kb-069ba31903658442/1.0.0/391e9ca2838fa532d14581d6ba6a63062b6fb22b50033f153ab9ddf0c4780cb7. Subsequent calls will reuse this data.
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 156.35it/s]
INFO:__main__:Checking global ID uniqueness
INFO:__main__:Found 38780 unique IDs
INFO:__main__:Gathering schema statistics
INFO:__main__:Gathering schema statistics
train
==========
id: 303
document_id: 303
passages: 303
entities: 23012
normalized: 0
events: 0
coreferences: 0
relations: 36348

test
==========
id: 202
document_id: 202
passages: 202
entities: 14914
normalized: 0
events: 0
coreferences: 0
relations: 23462

INFO:__main__:Checking if referenced IDs are properly mapped
INFO:__main__:KB ONLY: Checking passage offsets
INFO:__main__:KB ONLY: Checking entity offsets
INFO:__main__:KB ONLY: Checking event offsets
INFO:__main__:KB ONLY: Checking coref offsets
.
----------------------------------------------------------------------
Ran 1 test in 17.172s

OK

```

### CODE REVIEW

**Comment by galtay** [comment]
hi @singhay  thanks for this contribution. I'm getting a lot of warnings like the following in my local test run, 
```
WARNING:__main__:Referenced element ('T27', 'entity') could not be found in existing ids
```
can you make sure you are using the most recent version of the testing script please.

File: biodatasets/n2c2_2018/n2c2_2018_track2.py [hunk 1]
```diff
@@ -0,0 +1,427 @@
+# coding=utf-8
```
Comment (galtay) [inline biodatasets/n2c2_2018/n2c2_2018_track2.py:1]: can we make the directory name match the dataset name to be consistent with the other dataset loaders? 
`biodatasets/n2c2_2018_track2/n2c2_2018_track2.py`

**Review by galtay — CHANGES_REQUESTED** [review]
some renaming and check for updated test script please

File: biodatasets/n2c2_2018/n2c2_2018_track2.py (hunk 1 above)
Comment (singhay) [reply biodatasets/n2c2_2018/n2c2_2018_track2.py:1]: I intentionally did not do that to accomodate for #490 so both tracks come under single directory. It works well with tests as well.

**Comment by singhay** [comment]
> hi @singhay thanks for this contribution. I'm getting a lot of warnings like the following in my local test run,
> 
> ```
> WARNING:__main__:Referenced element ('T27', 'entity') could not be found in existing ids
> ```
> 
> can you make sure you are using the most recent version of the testing script please.

@galtay thank you for pointing this out, I was indeed was previous version of test scripts.

I was able to reproduce the warning messages and my latest commit ✅ the 🐛 . 
Thank you 🙏🏼 for taking the time.

File: biodatasets/n2c2_2018/n2c2_2018_track2.py (hunk 1 above)
Comment (galtay) [reply biodatasets/n2c2_2018/n2c2_2018_track2.py:1]: yes, I understand. I think it's better to follow the pattern that already exists in the repository though. for example in biodatasets, 
```
├── tmvar_v1
│   └── tmvar_v1.py
├── tmvar_v2
│   └── tmvar_v2.py
```

```
├── n2c2_2006_deid
│   └── n2c2_2006_deid.py
├── n2c2_2006_smokers
│   └── n2c2_2006_smokers.py
```

```
├── bioasq_task_b
│   └── bioasq_task_b.py
├── bioasq_task_c_2017
│   └── bioasq_task_c_2017.py
```

Comment (singhay) [reply biodatasets/n2c2_2018/n2c2_2018_track2.py]: Thanks, I've updated as suggested.

**Comment by singhay** [comment]
Closes #37

File: biodatasets/n2c2_2018_track2/n2c2_2018_track2.py [hunk 2]
```diff
@@ -0,0 +1,436 @@
+# coding=utf-8
+# Copyright 2022 The HuggingFace Datasets Authors and Ayush Singh (singhay).
```
Comment (galtay) [inline biodatasets/n2c2_2018_track2/n2c2_2018_track2.py:2]: initially the template requested that participants add names here, but then we decided to use github to track contributions. can you make this consistent with the template please https://github.com/bigscience-workshop/biomedical/blob/master/templates/template.py#L2

File: biodatasets/n2c2_2018_track2/n2c2_2018_track2.py [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+    BIGBIO_VERSION = datasets.Version(_BIGBIO_VERSION)
+
+    # You will be able to load the "source" or "bigbio" configurations with
+    # ds_source = datasets.load_dataset('my_dataset', name='source')
+    # ds_bigbio = datasets.load_dataset('my_dataset', name='bigbio')
+
+    # For local datasets you can make use of the `data_dir` and `data_files` kwargs
+    # https://huggingface.co/docs/datasets/add_dataset.html#downloading-data-files-and-organizing-splits
+    # ds_source = datasets.load_dataset('my_dataset', name='source', data_dir="/path/to/data/files")
+    # ds_bigbio = datasets.load_dataset('my_dataset', name='bigbio', data_dir="/path/to/data/files")
```
Comment (galtay) [inline biodatasets/n2c2_2018_track2/n2c2_2018_track2.py:301]: lets remove these outdated comments please (remove line 294-301)

File: biodatasets/n2c2_2018_track2/n2c2_2018_track2.py [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+
+    def _split_generators(self, dl_manager) -> List[datasets.SplitGenerator]:
+        """Returns SplitGenerators."""
+        if self.config.data_dir is None or self.config.name is None:
+            raise ValueError("This is a local dataset. Please pass the data_dir and name kwarg to load_dataset.")
+        else:
+            data_dir = self.config.data_dir
+
+        # Not all datasets have predefined canonical train/val/test splits.
+        # If your dataset has no predefined splits, use datasets.Split.TRAIN for all of the data.
```
Comment (galtay) [inline biodatasets/n2c2_2018_track2/n2c2_2018_track2.py:371]: please remove comments here

**Review by galtay — CHANGES_REQUESTED** [review]
just a few cosmetic changes and I think this is ready to merge. thanks for the contribution!

File: biodatasets/n2c2_2018_track2/n2c2_2018_track2.py (hunk 2 above)
Comment (singhay) [reply biodatasets/n2c2_2018_track2/n2c2_2018_track2.py]: done

File: biodatasets/n2c2_2018_track2/n2c2_2018_track2.py (hunk 3 above)
Comment (singhay) [reply biodatasets/n2c2_2018_track2/n2c2_2018_track2.py]: done

File: biodatasets/n2c2_2018_track2/n2c2_2018_track2.py (hunk 4 above)
Comment (singhay) [reply biodatasets/n2c2_2018_track2/n2c2_2018_track2.py]: done

**Review by galtay — APPROVED** [review]

**Pull request merged**